File: examples/asyncapi/provider/src/main/java/io/example/asyncapi/provider/service/impl/CustomerServiceImpl.java
package io.example.asyncapi.provider.service.impl;

import io.example.asyncapi.provider.domain.*;
import io.example.asyncapi.provider.repository.jpa.*;
import io.example.asyncapi.provider.service.*;
import io.example.asyncapi.provider.service.dtos.*;
import io.example.asyncapi.provider.service.impl.mappers.*;
import java.math.*;
import java.time.*;
import java.util.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service Implementation for managing [Customer].
 */
@Service
@Transactional(readOnly = true)
@lombok.AllArgsConstructor
public class CustomerServiceImpl implements CustomerService {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private final CustomerServiceMapper customerServiceMapper = CustomerServiceMapper.INSTANCE;

    private final CustomerRepository customerRepository;

    @Transactional
    public Customer createCustomer(Customer input) {
        log.debug("[CRUD] Request to save Customer: {}", input);
        var customer = customerServiceMapper.update(new Customer(), input);
        customer = customerRepository.save(customer);
        // TODO: may need to reload the entity to fetch relationships 'mapped by id'
        return customer;
    }

    public Optional<Customer> getCustomer(Long id) {
        log.debug("[CRUD] Request to get Customer : {}", id);
        var customer = customerRepository.findById(id);
        return customer;
    }

    @Transactional
    public Optional<Customer> updateCustomer(Long id, Customer input) {
        log.debug("Request updateCustomer: {} {}", id, input);

        var customer = customerRepository
                .findById(id)
                .map(existingCustomer -> {
                    return customerServiceMapper.update(existingCustomer, input);
                })
                .map(customerRepository::save);
        return customer;
    }

    @Transactional
    public void deleteCustomer(Long id) {
        log.debug("[CRUD] Request to delete Customer : {}", id);
        customerRepository.deleteById(id);
    }

    public Page<Customer> listCustomers(Pageable pageable) {
        log.debug("Request listCustomers: {}", pageable);

        var customers = customerRepository.findAll(pageable);
        return customers;
    }
}

File: examples/kustomer-address-jpa/src/main/kotlin/io/zenwave360/examples/kotlin/service/impl/CustomerServiceImpl.kt
package io.zenwave360.examples.kotlin.service.impl

import io.zenwave360.examples.kotlin.domain.*
import io.zenwave360.examples.kotlin.service.*
import io.zenwave360.examples.kotlin.service.dtos.*
import io.zenwave360.examples.kotlin.service.impl.mappers.*
import io.zenwave360.examples.kotlin.repository.jpa.*
// import io.zenwave360.examples.kotlin.events.dtos.*
import io.zenwave360.examples.kotlin.events.*

import java.math.*
import java.time.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.Optional

/**
 * Service Implementation for managing [Customer].
 */
@Service
@Transactional(readOnly = true)
open class CustomerServiceImpl(

    private val customerRepository: CustomerRepository


    , private val eventsProducer: CustomerEventsProducer

) : CustomerService {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    private val customerServiceMapper: CustomerServiceMapper = CustomerServiceMapper.INSTANCE


    private val eventsMapper: EventsMapper = EventsMapper.INSTANCE



    @Transactional

    override  fun createCustomer(input: Customer): Customer
 {
    log.debug("[CRUD] Request to save Customer: {}", input)
    var customer = customerServiceMapper.update(Customer(), input)
    customer = customerRepository.save(customer)
    // TODO: may need to reload the entity to fetch relationships 'mapped by id'
    
        // emit events
            val customerEvent = eventsMapper.asCustomerEvent(customer)
            eventsProducer.onCustomerEvent(customerEvent)
    return customer


}


    override  fun getCustomer(id: Long): Optional<Customer>
 {
    log.debug("[CRUD] Request to get Customer : {}", id)
    val customer = customerRepository.findById(id)
    return customer


}

    @Transactional

    override  fun updateCustomer(id: Long, input: Customer): Optional<Customer>
 {
    log.debug("Request updateCustomer: {} {}", id, input);

    val customer = customerRepository.findById(id).map { existingCustomer ->
        customerServiceMapper.update(existingCustomer, input)
    }
    .map { customerRepository.save(it) }
     if(customer.isPresent()) {
        // emit events
            val customerEvent = eventsMapper.asCustomerEvent(customer.get())
            eventsProducer.onCustomerEvent(customerEvent) }
    return customer


}

    @Transactional

    override  fun deleteCustomer(id: Long): Unit
 {
    log.debug("[CRUD] Request to delete Customer : {}", id)
    customerRepository.deleteById(id)
    
        // emit events
            val customerEvent = eventsMapper.asCustomerEvent(id)
            eventsProducer.onCustomerEvent(customerEvent)


}


    override  fun searchCustomers(input: CustomerSearchCriteria, pageable: Pageable): Page<Customer>
 {
    log.debug("Request searchCustomers: {} {}", input, pageable);

    val customers = customerRepository.findAll(pageable)
    return customers


}



}
